File: app/src/main/java/com/example/data/model/UserProfile.kt
package com.example.data.model

import com.google.firebase.database.IgnoreExtraProperties
import com.google.firebase.database.PropertyName

/**
 * BalanceCache model matching Firebase Realtime Database schema:
 * { "balance": 0.3923, "currency": "USD", "updatedAt": 1786685279 }
 */
@IgnoreExtraProperties
data class BalanceCache(
    var balance: Double = 0.00,
    var currency: String = "USD",
    var updatedAt: Long = System.currentTimeMillis()
)

/**
 * SipConfig for the local SIP stack. The password field is local-only and must
 * never be written to Firebase RTDB.
 */
@IgnoreExtraProperties
data class SipConfig(
    var callerId: String = "",
    var deviceId: String = "",
    var host: String = "",
    var password: String = "",
    var port: Int = 5060,
    var username: String = "",
    var updatedAt: Long = System.currentTimeMillis()
) {
    fun hasUsableCredentials(): Boolean {
        return host.isNotBlank() && username.isNotBlank() && password.isNotBlank()
    }

    fun registrationFingerprint(): String {
        return "$host|$port|$username|$password"
    }

    fun toRemoteMap(): Map<String, Any?> {
        return mapOf(
            "callerId" to callerId,
            "deviceId" to deviceId,
            "host" to host,
            "port" to port,
            "username" to username,
            "updatedAt" to updatedAt
        )
    }
}

/**
 * UserProfile data model representing real true user identity,
 * credit balance, verified caller ID, SIP trunk configuration, and real-time synchronization
 * with Firebase Realtime Database.
 */
@IgnoreExtraProperties
data class UserProfile(
    @get:PropertyName("uid")
    @set:PropertyName("uid")
    var uid: String = GUEST_UID,

    @get:PropertyName("displayName")
    @set:PropertyName("displayName")
    var displayName: String = "Guest Operator",

    @get:PropertyName("email")
    @set:PropertyName("email")
    var email: String = "operator@dialerid.secure",

    @get:PropertyName("phoneNumber")
    @set:PropertyName("phoneNumber")
    var phoneNumber: String = "",

    @get:PropertyName("photoUrl")
    @set:PropertyName("photoUrl")
    var photoUrl: String? = null,

    @get:PropertyName("creditBalance")
    @set:PropertyName("creditBalance")
    var creditBalance: Double = 0.00,

    @get:PropertyName("currency")
    @set:PropertyName("currency")
    var currency: String = "USD",

    @get:PropertyName("selectedCallerId")
    @set:PropertyName("selectedCallerId")
    var selectedCallerId: String = "",

    @get:PropertyName("accountType")
    @set:PropertyName("accountType")
    var accountType: String = "Enterprise VoIP Trunk",

    @get:PropertyName("accountRole")
    @set:PropertyName("accountRole")
    var accountRole: String = "Verified Operator",

    @get:PropertyName("organization")
    @set:PropertyName("organization")
    var organization: String = "Secure Telecom Network",

    @get:PropertyName("presence")
    @set:PropertyName("presence")
    var presence: String = "Online & Ready",

    @get:PropertyName("networkStatus")
    @set:PropertyName("networkStatus")
    var networkStatus: String = "Realtime DB Connected",

    @get:PropertyName("isVerified")
    @set:PropertyName("isVerified")
    var isVerified: Boolean = false,

    @get:PropertyName("isEncrypted")
    @set:PropertyName("isEncrypted")
    var isEncrypted: Boolean = false,

    @get:PropertyName("audioQuality")
    @set:PropertyName("audioQuality")
    var audioQuality: String = "Standard Line",

    @get:PropertyName("preferredCodec")
    @set:PropertyName("preferredCodec")
    var preferredCodec: String = "G711_AUTO",

    @get:PropertyName("callsCount")
    @set:PropertyName("callsCount")
    var callsCount: Int = 0,

    @get:PropertyName("totalMinutes")
    @set:PropertyName("totalMinutes")
    var totalMinutes: Int = 0,

    @get:PropertyName("sipConfig")
    @set:PropertyName("sipConfig")
    var sipConfig: SipConfig? = null,

    @get:PropertyName("isCloudSynced")
    @set:PropertyName("isCloudSynced")
    var isCloudSynced: Boolean = false,

    @get:PropertyName("createdAt")
    @set:PropertyName("createdAt")
    var createdAt: Long = System.currentTimeMillis(),

    @get:PropertyName("lastSyncTimestamp")
    @set:PropertyName("lastSyncTimestamp")
    var lastSyncTimestamp: Long = 0L,

    @get:PropertyName("lastUpdated")
    @set:PropertyName("lastUpdated")
    var lastUpdated: Long = System.currentTimeMillis()
) {
    val isGuest: Boolean
        get() = uid.isBlank() || uid == GUEST_UID
    /**
     * Converts this UserProfile instance to a Map conforming to Firebase Realtime Database schema.
     */
    fun toMap(): Map<String, Any?> {
        return Companion.toMap(this)
    }

    companion object {
        const val GUEST_UID = "guest_operator_001"
        /**
         * Serializes a UserProfile object into a Map for Realtime Database write/update operations.
         */
        fun toMap(profile: UserProfile): Map<String, Any?> {
            val profileMap = mapOf(
                "display_name" to profile.displayName,
                "displayName" to profile.displayName,
                "email" to profile.email,
                "createdAt" to profile.createdAt
            )

            val balanceCacheMap = mapOf(
                "balance" to profile.creditBalance,
                "currency" to profile.currency,
                "updatedAt" to System.currentTimeMillis()
            )

            val rootMap = mutableMapOf<String, Any?>(
                "uid" to profile.uid,
                "profile" to profileMap,
                "balanceCache" to balanceCacheMap,
                "selectedCallerId" to profile.selectedCallerId,
                "displayName" to profile.displayName,
                "email" to profile.email,
                "phoneNumber" to profile.phoneNumber,
                "photoUrl" to profile.photoUrl,
                "creditBalance" to profile.creditBalance,
                "accountType" to profile.accountType,
                "accountRole" to profile.accountRole,
                "organization" to profile.organization,
                "presence" to profile.presence,
                "networkStatus" to profile.networkStatus,
                "isVerified" to profile.isVerified,
                "isEncrypted" to profile.isEncrypted,
                "audioQuality" to profile.audioQuality,
                "preferredCodec" to profile.preferredCodec,
                "callsCount" to profile.callsCount,
                "totalMinutes" to profile.totalMinutes,
                "isCloudSynced" to profile.isCloudSynced,
                "createdAt" to profile.createdAt,
                "lastSyncTimestamp" to profile.lastSyncTimestamp,
                "lastUpdated" to System.currentTimeMillis()
            )

            profile.sipConfig?.let { sip ->
                rootMap["sip"] = sip.toRemoteMap()
            }

            return rootMap
        }

        /**
         * Deserializes a Realtime Database snapshot map dynamically into a UserProfile instance,
         * gracefully parsing nested `profile`, `balanceCache`, `sip`, and top-level fields.
         */
        @Suppress("UNCHECKED_CAST")
        fun fromMap(map: Map<String, Any?>, uid: String = ""): UserProfile {
            val resolvedUid = uid.ifBlank { (map["uid"] as? String) ?: GUEST_UID }

            // Extract nested profile object if present
            val nestedProfile = map["profile"] as? Map<String, Any?>

            val displayName = (nestedProfile?.get("display_name") as? String)
                ?: (nestedProfile?.get("displayName") as? String)
                ?: (map["displayName"] as? String)
                ?: (map["display_name"] as? String)
                ?: "Operator"

            val email = (nestedProfile?.get("email") as? String)
                ?: (map["email"] as? String)
                ?: "operator@dialerid.secure"

            val rawCreatedAt = (nestedProfile?.get("createdAt") as? Number)?.toLong()
                ?: (map["createdAt"] as? Number)?.toLong()
                ?: System.currentTimeMillis()
            // If createdAt is in unix seconds (< 10000000000), convert to millis
            val createdAt = if (rawCreatedAt in 1..9999999999L) rawCreatedAt * 1000L else rawCreatedAt

            // Extract nested balanceCache object if present
            val nestedBalance = map["balanceCache"] as? Map<String, Any?>
            val rawBalance = (nestedBalance?.get("balance"))
                ?: map["creditBalance"]
                ?: map["balance"]

            val balance = when (rawBalance) {
                is Number -> rawBalance.toDouble()
                is String -> rawBalance.toDoubleOrNull() ?: 0.00
                else -> 0.00
            }

            val currency = (nestedBalance?.get("currency") as? String)
                ?: (map["currency"] as? String)
                ?: "USD"

            val selectedCallerId = (map["selectedCallerId"] as? String)
                ?: (map["selected_caller_id"] as? String)
                ?: ""

            // Extract SIP configuration if present
            val nestedSip = map["sip"] as? Map<String, Any?>
            val sipConfig = if (nestedSip != null) {
                val portVal = when (val p = nestedSip["port"]) {
                    is Number -> p.toInt()
                    is String -> p.toIntOrNull() ?: 5060
                    else -> 5060
                }
                SipConfig(
                    callerId = (nestedSip["callerId"] as? String) ?: selectedCallerId,
                    deviceId = (nestedSip["deviceId"]?.toString()) ?: "",
                    host = (nestedSip["host"] as? String) ?: "",
                    password = "",
                    port = portVal,
                    username = (nestedSip["username"] as? String) ?: "",
                    updatedAt = (nestedSip["updatedAt"] as? Number)?.toLong() ?: System.currentTimeMillis()
                )
            } else null

            val calls = when (val value = map["callsCount"]) {
                is Number -> value.toInt()
                is String -> value.toIntOrNull() ?: 0
                else -> 0
            }

            val minutes = when (val value = map["totalMinutes"]) {
                is Number -> value.toInt()
                is String -> value.toIntOrNull() ?: 0
                else -> 0
            }

            return UserProfile(
                uid = resolvedUid,
                displayName = displayName,
                email = email,
                phoneNumber = (map["phoneNumber"] as? String) ?: "",
                photoUrl = map["photoUrl"] as? String,
                creditBalance = balance,
                currency = currency,
                selectedCallerId = selectedCallerId,
                accountType = (map["accountType"] as? String) ?: "Enterprise VoIP Trunk",
                accountRole = (map["accountRole"] as? String) ?: "Verified Operator",
                organization = (map["organization"] as? String) ?: "Secure Telecom Network",
                presence = (map["presence"] as? String) ?: "Online & Ready",
                networkStatus = (map["networkStatus"] as? String) ?: "Realtime DB Synced",
                isVerified = (map["isVerified"] as? Boolean) ?: false,
                isEncrypted = (map["isEncrypted"] as? Boolean) ?: false,
                audioQuality = (map["audioQuality"] as? String) ?: "Standard Line",
                preferredCodec = (map["preferredCodec"] as? String) ?: "G711_AUTO",
                callsCount = calls,
                totalMinutes = minutes,
                sipConfig = sipConfig,
                isCloudSynced = true,
                createdAt = createdAt,
                lastSyncTimestamp = System.currentTimeMillis(),
                lastUpdated = (map["lastUpdated"] as? Number)?.toLong() ?: System.currentTimeMillis()
            )
        }
    }
}

File: ios/DialerID/Data/Models/UserProfile.swift
import Foundation

struct BalanceCache: Equatable {
    var balance: Double = 0
    var currency: String = "USD"
    var updatedAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
}

struct SipConfig: Equatable {
    var callerId: String = ""
    var deviceId: String = ""
    var host: String = ""
    var password: String = ""
    var port: Int = 5060
    var username: String = ""
    var updatedAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000)

    func hasIdentity() -> Bool {
        !host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func needsPassword() -> Bool {
        hasIdentity() && password.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func hasUsableCredentials() -> Bool {
        hasIdentity() && !password.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func registrationFingerprint() -> String {
        "\(host)|\(port)|\(username)|\(password)|\(deviceId)"
    }

    func withResolvedPassword(_ localPassword: String) -> SipConfig {
        let resolved = password.isEmpty ? localPassword : password
        if resolved == password { return self }
        var copy = self
        copy.password = resolved
        return copy
    }

    func toRemoteMap() -> [String: Any] {
        [
            "callerId": callerId,
            "deviceId": deviceId,
            "host": host,
            "password": password,
            "port": port,
            "username": username,
            "updatedAt": updatedAt
        ]
    }
}

struct UserProfile: Equatable {
    static let guestUID = "guest_operator_001"

    var uid: String = guestUID
    var displayName: String = "Guest Operator"
    var email: String = "operator@dialerid.secure"
    var phoneNumber: String = ""
    var photoUrl: String?
    var creditBalance: Double = 0
    var currency: String = "USD"
    var selectedCallerId: String = ""
    var accountType: String = "Enterprise VoIP Trunk"
    var accountRole: String = ""
    var organization: String = ""
    var presence: String = "Online & Ready"
    var networkStatus: String = "Realtime DB Connected"
    var isVerified = false
    var isEncrypted = false
    var audioQuality: String = "Standard Line"
    var preferredCodec: String = "G711_AUTO"
    var callsCount: Int = 0
    var totalMinutes: Int = 0
    var assignedSipId: String = ""
    var sipConfig: SipConfig?
    var isCloudSynced = false
    var createdAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    var lastSyncTimestamp: Int64 = 0
    var lastUpdated: Int64 = Int64(Date().timeIntervalSince1970 * 1000)

    var isGuest: Bool {
        uid.isEmpty || uid == Self.guestUID
    }

    func toMap() -> [String: Any] {
        var root: [String: Any] = [
            "uid": uid,
            "profile": [
                "display_name": displayName,
                "displayName": displayName,
                "email": email,
                "createdAt": createdAt
            ],
            "balanceCache": [
                "balance": creditBalance,
                "currency": currency,
                "updatedAt": Int64(Date().timeIntervalSince1970 * 1000)
            ],
            "selectedCallerId": selectedCallerId,
            "displayName": displayName,
            "email": email,
            "phoneNumber": phoneNumber,
            "photoUrl": photoUrl as Any,
            "creditBalance": creditBalance,
            "accountType": accountType,
            "accountRole": accountRole,
            "organization": organization,
            "presence": presence,
            "networkStatus": networkStatus,
            "isVerified": isVerified,
            "isEncrypted": isEncrypted,
            "audioQuality": audioQuality,
            "preferredCodec": preferredCodec,
            "callsCount": callsCount,
            "totalMinutes": totalMinutes,
            "isCloudSynced": isCloudSynced,
            "createdAt": createdAt,
            "lastSyncTimestamp": lastSyncTimestamp,
            "lastUpdated": Int64(Date().timeIntervalSince1970 * 1000),
            "assignedSipId": assignedSipId
        ]
        if let deviceId = sipConfig?.deviceId, !deviceId.isEmpty {
            root["deviceId"] = deviceId
        }
        if let sip = sipConfig {
            root["sip"] = sip.toRemoteMap()
        }
        return root
    }

    static func fromMap(_ map: [String: Any], uid: String = "") -> UserProfile {
        let resolvedUid: String
        if !uid.isEmpty {
            resolvedUid = uid
        } else if let value = map["uid"] as? String {
            resolvedUid = value
        } else {
            resolvedUid = guestUID
        }

        let nestedProfile = map["profile"] as? [String: Any]
        let displayName = (nestedProfile?["display_name"] as? String)
            ?? (nestedProfile?["displayName"] as? String)
            ?? (map["displayName"] as? String)
            ?? (map["display_name"] as? String)
            ?? "Operator"
        let email = (nestedProfile?["email"] as? String)
            ?? (map["email"] as? String)
            ?? "operator@dialerid.secure"
        var rawCreatedAt = int64(nestedProfile?["createdAt"]) ?? int64(map["createdAt"])
            ?? Int64(Date().timeIntervalSince1970 * 1000)
        if (1...9_999_999_999).contains(rawCreatedAt) {
            rawCreatedAt *= 1000
        }

        let nestedBalance = map["balanceCache"] as? [String: Any]
        let rawBalance = nestedBalance?["balance"] ?? map["creditBalance"] ?? map["balance"]
        let balance = double(rawBalance)
        let currency = (nestedBalance?["currency"] as? String) ?? (map["currency"] as? String) ?? "USD"
        let selectedCallerId = (map["selectedCallerId"] as? String)
            ?? (map["selected_caller_id"] as? String)
            ?? ""
        let assignedSipId = (map["assignedSipId"] as? String)
            ?? (map["assigned_sip_id"] as? String)
            ?? ""
        let rootDeviceId = "\(map["deviceId"] ?? "")"

        var sipConfig: SipConfig?
        if let nestedSip = map["sip"] as? [String: Any] {
            let port: Int
            if let number = nestedSip["port"] as? NSNumber {
                port = number.intValue
            } else if let string = nestedSip["port"] as? String {
                port = Int(string) ?? 5060
            } else {
                port = 5060
            }
            let password: String
            if let string = nestedSip["password"] as? String {
                password = string
            } else if let number = nestedSip["password"] as? NSNumber {
                password = number.stringValue
            } else {
                password = ""
            }
            sipConfig = SipConfig(
                callerId: (nestedSip["callerId"] as? String) ?? selectedCallerId,
                deviceId: {
                    let value = "\(nestedSip["deviceId"] ?? "")"
                    return value.isEmpty ? rootDeviceId : value
                }(),
                host: nestedSip["host"] as? String ?? "",
                password: password,
                port: port,
                username: {
                    let value = nestedSip["username"] as? String ?? ""
                    return value.isEmpty ? assignedSipId : value
                }(),
                updatedAt: int64(nestedSip["updatedAt"]) ?? Int64(Date().timeIntervalSince1970 * 1000)
            )
        } else if !assignedSipId.isEmpty || !rootDeviceId.isEmpty {
            sipConfig = SipConfig(
                deviceId: rootDeviceId,
                username: assignedSipId,
                updatedAt: Int64(Date().timeIntervalSince1970 * 1000)
            )
        }

        return UserProfile(
            uid: resolvedUid,
            displayName: displayName,
            email: email,
            phoneNumber: map["phoneNumber"] as? String ?? "",
            photoUrl: map["photoUrl"] as? String,
            creditBalance: balance,
            currency: currency,
            selectedCallerId: selectedCallerId,
            accountType: map["accountType"] as? String ?? "Enterprise VoIP Trunk",
            accountRole: map["accountRole"] as? String ?? "",
            organization: map["organization"] as? String ?? "",
            presence: map["presence"] as? String ?? "Online & Ready",
            networkStatus: map["networkStatus"] as? String ?? "Realtime DB Synced",
            isVerified: map["isVerified"] as? Bool ?? false,
            isEncrypted: map["isEncrypted"] as? Bool ?? false,
            audioQuality: map["audioQuality"] as? String ?? "Standard Line",
            preferredCodec: map["preferredCodec"] as? String ?? "G711_AUTO",
            callsCount: int(map["callsCount"]),
            totalMinutes: int(map["totalMinutes"]),
            assignedSipId: assignedSipId,
            sipConfig: sipConfig,
            isCloudSynced: true,
            createdAt: rawCreatedAt,
            lastSyncTimestamp: Int64(Date().timeIntervalSince1970 * 1000),
            lastUpdated: int64(map["lastUpdated"]) ?? Int64(Date().timeIntervalSince1970 * 1000)
        )
    }

    private static func double(_ value: Any?) -> Double {
        switch value {
        case let number as NSNumber: return number.doubleValue
        case let string as String: return Double(string) ?? 0
        default: return 0
        }
    }

    private static func int(_ value: Any?) -> Int {
        switch value {
        case let number as NSNumber: return number.intValue
        case let string as String: return Int(string) ?? 0
        default: return 0
        }
    }

    private static func int64(_ value: Any?) -> Int64? {
        switch value {
        case let number as NSNumber: return number.int64Value
        case let string as String: return Int64(string)
        default: return nil
        }
    }
}

enum CallStatus: String, Codable, CaseIterable {
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"
    case failed = "FAILED"
    case noAnswer = "NO_ANSWER"
    case disconnected = "DISCONNECTED"
}

struct CallerIdItem: Equatable, Identifiable {
    var id: String
    var phoneNumber: String
    var label: String
    var isPrimary = false
    var isVerified = false
    var countryCode: String = "US"
    var host: String = ""
    var port: String = ""
    var username: String = ""
    var createdAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
}

struct CallLogItem: Equatable, Identifiable {
    var id: String
    var destinationNumber: String
    var callerIdUsed: String
    var countryName: String
    var status: CallStatus
    var timestamp: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    var durationSeconds: Int
    var billingRatePerMin: Double
    var totalCost: Double
}
